test/test_engine_struct.js

```diff
@@ -0,0 +1,626 @@
+        new EngineNumber(10, "tCO2e"),
+        new EngineNumber(2, "units"),
+      );
+    };
+
+    QUnit.test("initializes", function (assert) {
+      const supplement = makeExample();
+      assert.notDeepEqual(supplement, undefined);
+    });
+
+    QUnit.test("getInitialChargeValue", function (assert) {
+      const supplement = makeExample();
+      const value = supplement.getInitialChargeValue();
+      assert.closeTo(value.getValue(), 5, 0.0001);
+      assert.deepEqual(value.getUnits(), "kg");
+    });
+
+    QUnit.test("getInitialChargeConsumption", function (assert) {
+      const supplement = makeExample();
+      const consumption = supplement.getInitialChargeConsumption();
+      assert.closeTo(consumption.getValue(), 10, 0.0001);
+      assert.deepEqual(consumption.getUnits(), "tCO2e");
+    });
+
+    QUnit.test("getNewPopulation", function (assert) {
+      const supplement = makeExample();
+      const population = supplement.getNewPopulation();
+      assert.closeTo(population.getValue(), 2, 0.0001);
+      assert.deepEqual(population.getUnits(), "units");
+    });
+  });
+
+  QUnit.module("EngineResultBuilder", function () {
+    const makeExampleBuilder = () => {
+      const builder = new EngineResultBuilder();
+      const importSupplement = new ImportSupplement(
+        new EngineNumber(5, "kg"),
+        new EngineNumber(10, "tCO2e"),
+        new EngineNumber(2, "units"),
+      );
+
+      builder.setApplication("test app");
+      builder.setSubstance("test substance");
+      builder.setYear(2023);
+      builder.setManufactureValue(new EngineNumber(100, "kg"));
+      builder.setImportValue(new EngineNumber(50, "kg"));
+      builder.setRecycleValue(new EngineNumber(25, "kg"));
+      builder.setDomesticConsumptionValue(new EngineNumber(200, "tCO2e"));
+      builder.setImportConsumptionValue(new EngineNumber(100, "tCO2e"));
+      builder.setRecycleConsumptionValue(new EngineNumber(50, "tCO2e"));
+      builder.setPopulationValue(new EngineNumber(1000, "units"));
+      builder.setPopulationNew(new EngineNumber(100, "units"));
+      builder.setRechargeEmissions(new EngineNumber(300, "tCO2e"));
+      builder.setEolEmissions(new EngineNumber(150, "tCO2e"));
+      builder.setEnergyConsumption(new EngineNumber(500, "kWh"));
+      builder.setImportSupplement(importSupplement);
+
+      return builder;
+    };
+
+    QUnit.test("initializes", function (assert) {
+      const builder = new EngineResultBuilder();
+      assert.notDeepEqual(builder, undefined);
+    });
+
+    QUnit.test("builds complete result", function (assert) {
+      const builder = makeExampleBuilder();
+      const result = builder.build();
+      assert.notDeepEqual(result, undefined);
+      assert.deepEqual(result.getApplication(), "test app");
+      assert.deepEqual(result.getSubstance(), "test substance");
+      assert.equal(result.getYear(), 2023);
+    });
+
+    QUnit.test("fails on incomplete result", function (assert) {
```

Comment:
Please call this fails on an empty result. Then, please add an additional test which checks failure for an almost complete result. Specifically, please fill everything except EOL emissions. Then, please add an additional test which checks that the builder given everything but the import supplement fails on build.